### ISSUE TITLE
Ability to trigger script calls from animations

### ISSUE THREAD

**Issue by eightyeight** [GarageGames/Torque3D#1300]
As per [this resource](http://www.garagegames.com/community/resources/view/21460). I thought we'd had proposals or PRs for this before, but I can't find them with a search.

**Comment by Azaezel** [comment]
Just https://github.com/GarageGames/Torque3D/pull/682 , and that one is very specifically limited to triggering off of the end of a given sequence, and tying it to the name of the animation just played via an onAnimPlayed callback

**Comment by Azaezel** [comment]
For clarity, to use the example from @RichardRanft (and to summon him for his own thoughts on the matter) for ease of comparison, the setup for the already included one would be:

function blacksmithData::onAdd(%this,%obj)
{
	%obj.playthread(0,"Hammer");
}

function blacksmithData::onHammer(%this,%obj)
{
	%obj.playthread(0,"Hammer");
        ServerPlay3D("Hammerfall",%obj.getTransform());
}

**Comment by RichardRanft** [comment]
Wrong Richard - too many Dicks around here...

That said - if this could be tied to the specific sequence and not the object that would totally rock.  This way trigger 8 would not do the same thing for the snore animation and for the slap animation, but all objects with that sequence assigned would trigger the same thing at the same point (I think...).

**Comment by eightyeight** [comment]
Ah, @Azaezel that was the one I was thinking of. @RichardsGameStudio is the Richard who submitted the resource.

**Comment by Azaezel** [comment]
Right. sorry bout that. That being said, I could most assuredly see the merit in including both. Doesn't look at first blush like there would be conflicts either.

**Comment by RichardsGameStudio** [comment]
Hi guys, I've noticed you've mentioned me. How can I help you? I'm rather busy right now migrating my website to HTML5/CSS3, so not so much T3D, but it should be ready very soon...

**Comment by Azaezel** [comment]
Mostly a question of if you'd support inclusion of the resource @eightyeight linked up top, and any modifications you'd want to see of it, or the other related PR to make life easier for folks, since you've  been down this road yourself from a pragmatic standpoint.

**Comment by RichardsGameStudio** [comment]
I have no problem with inclusion and or modification. Reading through this, maybe you can add http://www.garagegames.com/community/resources/view/21654 too, which can give you the name of the current animation. I'm sorry, but I don't have time to combine them atm (I see this can improve it all), also I'm not that great on github, so if someone can add this stuff, go ahead.

**Comment by Alex-doc** [comment]
What about getting the name directly on the trigger called? similar to this here http://www.garagegames.com/community/resources/view/21460/1#comment-186479

so in script we will have something like
(trigger set to "ambient")

function myObject::onAmbientAnimationTrigger(%this,%obj,%trigger,%sequence)

with:
```cpp
			if (st.thread && mShapeInstance && !isGhost() && !st.atEnd)
			{
				for (U32 j = 1; j < 33; j++)
				{
					if (mShapeInstance->getTriggerState(j))
					{
						char slot[3];
						dSprintf(slot, sizeof(slot), "%d", j);
						Con::executef(mDataBlock, 
							//call the script with sequence name
							"on"+st.thread->getSequenceName()+"AnimationTrigger",
							//use it as argument just in case
							getIdString(), slot, st.thread->getSequenceName()); 
					}
				}
			}
```
then for getting the sequence names/threads something like:

```cpp
String ShapeBase::getThreadSequenceName(S32 thread)
{
	Thread& st = mScriptThread[thread];
	if (st.thread)
		return st.thread->getSequenceName();
	else
		return "";
}

S32 ShapeBase::getSequenceThreadNumber(String name)
{
	for (S32 i = 0; i < MaxScriptThreads; i++)
	{
		Thread& st = mScriptThread[i];
		if (st.thread && st.thread->getSequenceName() == name)
		{
			   return i;
		}
		else
			continue;
	}
	return -1;
}

```

and then exposing them to script.
